### PR TITLE
Refactor and fix allocFirst with buckets for RubyHash

### PR DESCRIPTION
**Pull request by ChrisBr**

### CODE REVIEW

**Comment by enebo** [comment]
@ChrisBr This all looked good at a glance but there are epic amounts of OOME in the travis build.  Could that be related to these changes or is travis acting up?

**Comment by ChrisBr** [comment]
> This all looked good at a glance but there are epic amounts of OOME in the travis build. Could that be related to these changes or is travis acting up?

Likely :thinking: I will have a look at it tonight

**Comment by ChrisBr** [comment]
Ok, I dropped a few suspicious commits, let's see if this solves it. I will after this PR is merged subsequently create new PRs with the dropped commits.

File: core/src/main/java/org/jruby/RubyHash.java [hunk 1]
```diff
@@ -369,17 +369,30 @@ private final void allocFirst() {
...
-            hashes = new int[buckets];
+            int nextPowOfTwo = nextPowOfTwo(buckets);
+            entries = new IRubyObject[nextPowOfTwo << 1];
+            bins = new int[nextPowOfTwo << 1];
+            hashes = new int[nextPowOfTwo];
             Arrays.fill(bins, EMPTY_BIN);
         }
     }
 
+    private final int nextPowOfTwo(final int i) {
```
Comment (ahorek) [inline core/src/main/java/org/jruby/RubyHash.java]: for inspiration https://github.com/jemalloc/jemalloc/pull/1303

**Comment by eregon** [comment]
Let's merge this, CI is green (except just one failure of ruby/spec which I will tag).

**Pull request merged**